Featherweight Async

LeanderK opened the pull request:
I've added a async-enumeration to the mathsat case. It's disabled per default but you can enable it by passing blocking=False.

Can someone verify that the tests work? I have some weird local errors about imports....

LeanderK committed:
we do not have team-wide vscode defaults

LeanderK committed:
added async enumeration

LeanderK committed:
run blocking per default

masinag commented on wmipa/enumeration/mathsat.py:
```diff
@@ -163,18 +164,26 @@ def _callback(model, converter, result):
+        #     solver.msat_env(),
+        #     [converter.convert(v) for v in atoms],
+        #     lambda model: _callback2(model, converter, models),
+        # )
```
I think this callback is not useful anymore

masinag commented on wmipa/enumeration/mathsat.py:
```diff
@@ -163,18 +164,26 @@ def _callback(model, converter, result):
-        for model in models:
+
+        def callback(model):
+            print("hi")
```
Remove the debug print :)

masinag commented on wmipa/enumeration/mathsat.py:
```diff
@@ -189,7 +198,94 @@ def _callback(model, converter, result):
+        #     solver.msat_env(),
+        #     [converter.convert(v) for v in atoms],
+        #     _callback3,
+        # )
```
Cleanup

masinag commented on wmipa/enumeration/mathsat.py:
```diff
@@ -189,7 +198,94 @@ def _callback(model, converter, result):
+                msat_env, atoms, converter, callback
+            )
+
+    def all_sat_stream(self, msat_env, atoms, converter, f) -> Generator:
```
- Should be private methods (`_` prefix)
- Please add proper typings (also return type). I think we should run mypy with flags `--disable-error-code=import-untyped --disallow-untyped-defs --disallow-incomplete-defs`

masinag commented on wmipa/enumeration/mathsat.py:
```diff
@@ -189,7 +198,94 @@ def _callback(model, converter, result):
+            converter: The converter to convert atoms to MathSAT format.
+            f: A function to apply to each model found.
+        """
+        q = queue.Queue()
```
mypy complains that `q` should be annotated

masinag reviewed (commented):
Looks good to me! I added some minors.

If you fix the mypy error, then you should see the tests passing in the CI (not sure why you have import errors locally, did you install the project with `pip install -e ".[test]"`?).

masinag commented:
@paolomorettin I don't know if we want the `blocking` option to be set by the user. If so, I think it should be set in the constructor. Also, both versions should be tested.

LeanderK commented:
yeah, some cleanup is needed :) Not sure whether we need the blocking option but I guess then we default to the async enumeration?

Both options should be already be tested in the enumeration-tests though

paolomorettin commented:
Hi @LeanderK , thanks for the PR! 

Before accepting the PR it should pass all tests.

Regarding the `blocking` argument, I have the feeling that it should be passed to the constructor, but I'm not sure.
- What are the pros/cons of the two approaches?

More importantly, shouldn't the non-blocking enumerator logic be somewhat agnostic from the actual enumeration algorithm? If so, we should instead consider defining it in a separate module.

masinag commented:
> More importantly, shouldn't the non-blocking enumerator logic be somewhat agnostic from the actual enumeration algorithm? If so, we should instead consider defining it in a separate module.

I think this that the way is implemented is dependant from the enumeration algorithm/solver API.
I.e., in this case we need the thread-based approach because the MathSAT API only accepts a callback. The z3 enumerator, instead, is already non-blocking.

paolomorettin commented:
> We could change the Enumerator from a Protocol to something explicitely inherited. Then we can push the logic in there?

I was thinking more in line with the wrapper-based approach that we used for integrators (see e.g. how `parallel-latte` and `latte` work in the command line interface `cli/cli.py`)

> If you mean an async-enumerator, then I also think it's doable and it makes sense. Moreover, with this approach one could set the Queue.maxsize attribute to guarantee that we don't use exponential space.

Right, an async wrapper was what I had in mind.

LeanderK committed:
unit tests

LeanderK committed:
cleanup

LeanderK commented:
I am not sure what you suggest @paolomorettin. We can not just add an async-wrapper to go from blocking enumeration to async, we need to decide this at the callback (once we're sequential, we stay sequential). But as @masinag said, the API-differs betwen Z3 and Mathsat.

LeanderK committed:
wrong type annotation

masinag commented:
For a non-blocking non-async MathSAT enumerator we can use the current implementation by initializing `q = Queue(maxsize=1)`.

For the async wrapper, perhaps we could have another thread-based solution, where the wrapper's `enumerate` function calls the wrapped `enumerate` in a separate thread, with another queue.

LeanderK commented:
ah, I think I know understand. So we could try to create a non-blocking version for both mathsat and z3 with the async-approach as a wrapper. The only (smaller) issue here is what to do with the two loops in case we have boolean variables. It's a bit of a detail really. We would then block the outer one until the inner one releases...not really an issue for me as I currently only have real variables.

LeanderK commented:
I am not sure how this could work as the blocking nature of the mathsat callback-style approach (`msat_all_sat` until we have finished) needs us to go straight to async in order to get an iterative, non-blocking style iteration.

masinag commented:
> The only (smaller) issue here is what to do with the two loops in case we have boolean variables. It's a bit of a detail really. We would then block the outer one until the inner one releases...not really an issue for me as I currently only have real variables.

Good point. I think this could be an intended behaviour, but we can always handle this (and make it configurable) in the `MathSATEnumerator` class.

> I am not sure how this could work as the blocking nature of the mathsat callback-style approach (msat_all_sat until we have finished) needs us to go straight to async in order to get an iterative, non-blocking style iteration.

I would use your async solution, but if we set `q.maxsize=1` then we only have one thread running at a time. 
In fact, `q.pop()` will block until the other thread calls `q.push()`, and vice-versa.

paolomorettin commented:
@masinag Exactly. 

Some context for @LeanderK : we refactored the integrators code to be more modular, moving away from the previous, sub-class heavy code. Integrators here are either:
- baseline integrators (e.g. LattE, rejection-based, ...)
- wrappers around other integrators (Parallel, Cache, AxisAligned, ...)

The second kind implements functionalities that are somewhat orthogonal and agnostic to the underlying integrator.
You can then have integrators like Cache(Parallel(LattE)) or stuff like that.

We want to apply a similar philosophy to enumerators, although currently we only have 2 baseline enumerators (`Z3Enumerator` and `MathSATEnumerator`, the former being the open-source default, the latter being the SOTA enumerator from our papers). 

Now in relation to this PR:

It took me a bit but now I understand why async would require a non-blocking enumerator anyway.

I was suggesting that async enumeration could be implemented as an enumeration wrapper, however, I wonder if it really make sense in the context of the standard `computeWMI` workflow, as it would require also an async integration pipeline to be useful.

On a second thought, maybe it simply makes more sense to offer non-blocking integrators as building blocks for different workflows / use cases (like PAL), and let the users implement async pipelines externally.

What do you think?

masinag commented:
> I was suggesting that async enumeration could be implemented as an enumeration wrapper, however, I wonder if it really make sense in the context of the standard `computeWMI` workflow, as it would require also an async integration pipeline to be useful.

I think async enumerator + sync integrator works, since the integrator runs in the main thread (possibly parallelizing to other processes), while the enumerator would look for truth assignments "in the background".
Maybe also the other way around would make sense (?).

> On a second thought, maybe it simply makes more sense to offer non-blocking integrators as building blocks for different workflows / use cases (like PAL), and let the users implement async pipelines externally.

Maybe, but I'm not sure how.

paolomorettin commented:
> > I was suggesting that async enumeration could be implemented as an enumeration wrapper, however, I wonder if it really make sense in the context of the standard `computeWMI` workflow, as it would require also an async integration pipeline to be useful.
> 
> I think async enumerator + sync integrator works, since the integrator runs in the main thread (possibly parallelizing to other processes), while the enumerator would look for truth assignments "in the background". Maybe also the other way around would make sense (?).
> 
> > On a second thought, maybe it simply makes more sense to offer non-blocking integrators as building blocks for different workflows / use cases (like PAL), and let the users implement async pipelines externally.
> 
> Maybe, but I'm not sure how.

What I mean is that obtaining the behaviour you suggest would require changing `computeWMI`, which currently first enumerates every integral, and only when it's done it calls `integrate_batch` (possibly in parallel). The method returns when every integral is computed.

Other pipelines are possible, e.g. having async enumeration and integration (like a producer-consumer architecture), but that's not necessarily the intended use.

Maybe the confusion comes from the current method's name `computeWMI`, which should instead reflect more the non-async nature of the pipeline? The distributed pipeline you suggest also computes a weighted model integral after all..

Does this clarify?

masinag commented:
Uh, yes, I see what you mean, I missed that. 
We should avoid calling `integrate_batch` and just call `integrate` for every enumerated sub-problem, and let the integrator "batch" problems if needed.

masinag commented:
As I was suggesting some time ago, we could call something like `integrator.tell_problem(...)` for each problem enumerated, and then `integrator.get_results()` to get the results when you need them.

Then the integrator can decide whether to immediately compute the result, wait for the `get_results` call, or compute them in the background.

paolomorettin commented:
> Uh, yes, I see what you mean, I missed that. We should avoid calling `integrate_batch` and just call `integrate` for every enumerated sub-problem, and let the integrator "batch" problems if needed.

The current implementation of `computeWMI` is implicitly based on the assumption that the enumeration is not the bottleneck. Therefore enumeration is done sequentially (in the "master") and then expensive integrations are done in parallel (by multiple workers).

If we simply call integrate (in the master), we lose parallelization on the integrals, and this is somehow accomodating a different assumption: that enumeration is slow and should proceed separately while the master consumes the produced integrals. 

In the producer-consumer architecture, the master is in charge of coordinating both an enumeration worker and possibly multiple integration workers, but that's a fairly more complicated architecture than the current `computeWMI`.

I am not sure if this is what we want in general, hence my suggestion that different kind of pipelines should be implemented in different methods, side-by-side with `computeWMI` (which should be given a different name).

The integrators / enumerators should be implemented in order to support these different pipelines / assumptions.

paolomorettin commented:
@LeanderK Exactly! There is no single use case, one could even be uninterested about the integration part, or viceversa, one could have a stream of convex integrals and not require enumeration. 
Or one could take the above producer-consumer architecture further, instead of returning a result once every integral is consumed, a process could read the current partial wmi and do something with that asynchronously..

It's a rabbit hole really :D

That said, on multicore architectures the most efficient way to approach exact WMI is leveraging multiprocessing to the fullest.

masinag commented:
I'm not sure how broad the range of scenarios you have in mind is. :)

I would say:
- an enumerator takes a support and a weight function and enumerates pairs `<convex polytope, integrand>`
- an integrator takes a (array of) such pairs and returns a (array of) float.

Does this exclude any scenarios? Maybe we could have different building blocks, such as samplers or other components.

Then we could provide different `Solvers` implementing some pre-defined algorithms.

LeanderK commented:
I think it's good to start with the non-blocking and with the async-wrapers for the enumerators. I think the integrators are less obvious, there's more choice here. I would try to cover a simple, definite integral, maybe parallelised over the cpu using LattE and focused on getting `computeWMI` to be a very straightforward (and at best line-by-line commented!) definition. Then, the user can take of on from there and the library can evolve with what's needed. Trying to do everything from the start can also result in a lot of bloat and technical debt.

LeanderK committed:
added async, non-blocking but sync mathsat and tests

LeanderK committed:
Merge remote-tracking branch 'upstream/featherweight' into featherweight_parallel

LeanderK committed:
bugfix

masinag commented:
> **test/unit/enumerators_test.py:43**
> 
>     * The AsyncEnumerator test case doesn't test with different max_queue_size values. Consider adding test cases with different queue sizes to ensure the async behavior works correctly under various configurations.
> 
> 
> ```
>         (AsyncEnumerator, {"enumerator": MathSATEnumerator()}),
> ```

I agree on this too, we can just try with one single constant value.

LeanderK committed:
fixed bug in tests, small improvements

LeanderK committed:
enumerator: maybe better test-bounds for the AsyncEnumerator Queue

LeanderK commented:
> > **test/unit/enumerators_test.py:43**
> > ```
> > * The AsyncEnumerator test case doesn't test with different max_queue_size values. Consider adding test cases with different queue sizes to ensure the async behavior works correctly under various configurations.
> > ```
> 
> I agree on this too, we can just try with one single constant value.

I've added it with two different values, 0 and 1. Funnily, they were not run before as `INSTANCES = product(SUPPORTS, WEIGHTS)` is slightly wrong, it's exhausted after one run and only the first enumerator was tested. I've changed by wrapping it with a list to allow for repeated iteration.

masinag commented on test/unit/enumerators_test.py:
```diff
@@ -35,25 +35,34 @@
+        (Z3Enumerator, {}),
+        (MathSATEnumerator, {"max_queue_size": 1}),
+        (MathSATEnumerator, {"max_queue_size": 0}),
+        (AsyncEnumerator, {"enumerator": MathSATEnumerator()}),
```
We should also test
```python
        (AsyncEnumerator, {"enumerator": Z3Enumerator()})
```

masinag commented on test/unit/enumerators_test.py:
```diff
@@ -34,26 +34,36 @@
+        (MathSATEnumerator, {"max_queue_size": 1}),
+        (MathSATEnumerator, {"max_queue_size": 0}),
+        (AsyncEnumerator, {"enumerator": MathSATEnumerator(), "max_queue_size": 0}),
+        (AsyncEnumerator, {"enumerator": MathSATEnumerator(), "max_queue_size": 1}),
```
We should also test
```python
        (AsyncEnumerator, {"enumerator": Z3Enumerator()})
```

LeanderK replied:
thought that's unnecessary, but easy fix

LeanderK committed:
added AsyncEnumerator/Z3Enumerator combination

LeanderK commented:
I think it's ok, and we can change it later. It's not entirely synchronous but I think it should still prevent things like mathsat hogging all the CPU, producing solutions like crazy while the consumer can't keep up as it can only produce the next in parallel.

paolomorettin commented:
> > I found some time to review the PR, good job!
> > One minor detail: I would rename `async_enumerator.py` to `async.py` to be consistent with the naming convention.
> 
> I know this is the naming convention used in the other wrappers, but `async` is a reserved keyword and I could not import the module with this name. This is how I ended up with `async_enumerator.py` as a quick fix. I am open to other names....`async_wrapper.py`? `asynchronous.py`?

Huh I didn't know that. Asynchronous is a bit of a mouthful, but I can't think of a better alternative.. @masinag ?

> Btw, I think this is not completely true, since you can have one thread searching for the next model while the other is consuming the previous one. You probably would need additional synchronization to get a real synchronous-like behavior, but I think we can keep it as it is now.

I agree, I don't think that losing fully synchronous behaviour is a huge loss. I wouldn't make the code overly complicated just for the sake of it.

masinag commented:
I would go for `asynchronous.py` or `async_.py` (trailing underscores are the [convention](https://peps.python.org/pep-0008/#descriptive-naming-styles) for names conflicting with Python keywords).

LeanderK committed:
enumerators: rename async_enumerator.py to asynchronous.py

LeanderK committed:
enumerators: rename AsyncEnumerator to AsyncWrapper

LeanderK commented:
I've opted for `asynchronous.py`....after all it's the import, we don't have to repeat this all the time. This way we can avoid the trailing underscore. Happy to change though

paolomorettin commented:
I think this is good, thanks again @LeanderK !
